File: crates/valang-runner/sdk/val.mjs
// The SDK a TypeScript or JavaScript handler is given.
//
// Verification already happened, in Rust, before this process started. There is
// one verifier and every language talks to it the same way: the runner hands
// the result in on stdin and this module shapes it.

export function sdk(checked) {
  return {
    verify: async () => checked,

    issuance(c, credential) {
      const e = (c.effects ?? []).find(
        (x) => x.capability === 'credential.issue' && x.payload?.credential === credential,
      )
      return e?.payload?.claims ?? null
    },

    issue: (credential, claims) => ({ kind: 'issue', credential, claims }),
    accept: (note) => ({ kind: 'accept', note }),
    refuse: (refusal) => ({ kind: 'refuse', refusal }),
  }
}

export async function main(handle) {
  const chunks = []
  for await (const chunk of process.stdin) chunks.push(chunk)
  const payload = JSON.parse(Buffer.concat(chunks).toString())
  const decision = await handle(payload.token, sdk(payload.checked))
  process.stdout.write(JSON.stringify(decision))
}

File: crates/valang-runner/sdk/val.rs
//! The SDK a Rust handler is given.
//!
//! Verification already happened, in the runner, before this process started.
//! There is one verifier and every language talks to it the same way: the
//! runner hands the result in on stdin and this module shapes it.

use serde_json::{json, Value};
use std::io::Read;

pub struct Checked {
    pub ok: bool,
    pub refusal: Value,
    pub record: Value,
    pub effects: Vec<Value>,
}

pub type Decision = Value;

pub struct Sdk {
    checked: Value,
}

impl Sdk {
    pub fn verify(&self, _token: &str) -> Result<Checked, Value> {
        let ok = self.checked["ok"].as_bool().unwrap_or(false);
        if !ok {
            return Err(self.checked["refusal"].clone());
        }
        Ok(Checked {
            ok,
            refusal: self.checked["refusal"].clone(),
            record: self.checked["record"].clone(),
            effects: self.checked["effects"].as_array().cloned().unwrap_or_default(),
        })
    }

    pub fn issuance(&self, checked: &Checked, credential: &str) -> Option<Value> {
        checked.effects.iter().find_map(|e| {
            if e["capability"] != "credential.issue" {
                return None;
            }
            if e["payload"]["credential"] == credential {
                Some(e["payload"]["claims"].clone())
            } else {
                None
            }
        })
    }

    pub fn issue(&self, credential: &str, claims: Value) -> Decision {
        json!({ "kind": "issue", "credential": credential, "claims": claims })
    }

    pub fn accept(&self, note: &str) -> Decision {
        json!({ "kind": "accept", "note": note })
    }

    pub fn refuse(&self, refusal: Value) -> Decision {
        json!({ "kind": "refuse", "refusal": refusal })
    }
}

pub fn main_with(handle: fn(&str, &Sdk) -> Decision) {
    let mut raw = String::new();
    std::io::stdin().read_to_string(&mut raw).expect("stdin");
    let payload: Value = serde_json::from_str(&raw).expect("json");
    let sdk = Sdk { checked: payload["checked"].clone() };
    let decision = handle(payload["token"].as_str().unwrap_or(""), &sdk);
    print!("{}", decision);
}
